forc-pkg/src/pkg.rs

```diff
@@ -193,6 +193,148 @@ impl BuildPlan {
+
+        // Collect dependency `Source`s from manifest.
+        let proj_id = self.graph[proj_node].id();
+        let proj_path = &self.path_map[&proj_id];
+        let manifest_dep_pkgs = manifest
+            .deps()
+            .map(|(dep_name, dep)| {
+                // NOTE: Temporarily warn about `version` until we have support for registries.
+                if let Dependency::Detailed(det) = dep {
+                    if det.version.is_some() {
+                        println_yellow_err(&format!(
+                            "  WARNING! Dependency \"{}\" specifies the unused `version` field: \
+                        consider using `branch` or `tag` instead",
+                            dep_name
+                        ));
+                    }
+                }
+
+                let name = dep.package().unwrap_or(dep_name).to_string();
+                let source = dep_to_source(proj_path, dep)?;
+                let dep_pkg = Pkg { name, source };
+                Ok((dep_name, dep_pkg))
+            })
+            .collect::<Result<BTreeSet<_>>>()?;
+        let added: Vec<Pkg> = manifest_dep_pkgs
+            .difference(&plan_dep_pkgs)
+            .clone()
+            .map(|pkg| (pkg.1.clone()))
+            .collect();
+        let removed: Vec<Pkg> = plan_dep_pkgs
+            .difference(&manifest_dep_pkgs)
+            .clone()
+            .map(|pkg| (pkg.1.clone()))
+            .collect();
+
+        Ok((added, removed))
+    }
+
+    pub fn from_old_manifest(
+        &mut self,
+        manifest_file: &ManifestFile,
+        sway_git_tag: &str,
+        offline_mode: bool,
+    ) -> Result<Self> {
+        let mut graph = self.graph.clone();
+        let proj_node = *self
+            .compilation_order
+            .last()
+            .ok_or_else(|| anyhow!("Invalid Graph"))?;
+        let proj_id = graph[proj_node].id();
+        let (added, removed) =
+            self.get_difference_with_manifest(proj_node, manifest_file, sway_git_tag)?;
+
+        use petgraph::visit::{Bfs, Walker};
+        //find root_node_id
+        let root_node_id = *self.compilation_order().last().unwrap();
+        let visited_nodes: Vec<NodeIx> = Bfs::new(&graph, root_node_id).iter(&graph).collect();
+        for node in visited_nodes {
+            if node == root_node_id {
+                continue;
+            }
+            // if a -> b in the graph, a depends on b. if a is removed b will have 0
+            // incoming edge, if there is no other parent dependening on b remove it.
+            // to remove a either it will be in removed list or it needs to have 0 incoming edge.
+            if graph.edges_directed(node, Direction::Incoming).count() == 0
+                || removed
+                    .clone()
+                    .into_iter()
+                    .any(|removed_dep| removed_dep.name == graph[node].name)
+            {
+                graph.remove_node(node);
+            }
+        }
```

Comment:
I was thinking on package removal a little more, and I think it might be more efficient and a bit cleaner if we start by removing the edges between the root and the removed nodes. I'm thinking something like this:

```rust
/// Remove the given set of packages from `graph` along with any dependencies that are no
/// longer required as a result.
fn remove_deps(
    graph: &mut Graph,
    path_map: &PathMap,
    proj_node: NodeIx,
    to_remove: &BTreeSet<Pkg>,
) {
    // Find the edges between the root and the removed packages.
    let edges_to_remove: Vec<_> = graph
        .edges_directed(proj_node, Direction::Outgoing)
        .filter_map(|e| {
            let dep_pkg = graph[e.target()].unpinned(path_map);
            if to_remove.contains(&dep_pkg) {
                Some(e.id())
            } else {
                None
            }
        })
        .collect();

    // Remove the edges.
    for e in edges_to_remove {
        graph.remove_edge(e);
    }

    // Do a BFS from the root and remove all nodes that are no longer connected to the root.
    let mut bfs = Bfs::new(graph, proj_node);
    bfs.next(graph); // Skip the root node (aka project node).
    while let Some(node) = bfs.next(graph) {
        if graph.edges_directed(node, Direction::Incoming).next().is_none() {
            graph.remove_node(node);
        }
    }
}
```